### ISSUE TITLE
"azure login -u" hangs with multi-factor auth

### ISSUE THREAD

**Issue by jinmiaoms** [Azure/azure-xplat-cli#2139]
I have CLI 0.9.9.  When I do "azure login -u", after entering password, it tells me to go to the website to enter code and do multi-factor auth.  I did that, but the command line just stays spinning without anything happening.  I'm running a Ubuntu VM in East US.

**Comment by yugangw-msft** [comment]
Could you please set the environment variable of "AZURE_ADAL_LOGGING_ENABLED" to 1, run the login, and share out the console log or send it to me at yugangw at microsoft dot com. Make sure to remove your password from the log.
Thanks for reporting this.
            Yugang

**Comment by squillace** [comment]
@jinmiaoms when you went to the web site, pasted in the code and cicked the Submit button, it takes you to a second page where it sorta implies you're logged in, but you're NOT. YOu have to continue and click the specific ID rectangle, typically the one that is highlighted, and then the cli will unblock. It should be only one step, but it's currently two.

I'm assuming you already figured this out..... or did you do that, and STILL hung....?

**Comment by mohitarora** [comment]
I have a separate issue. I get following error when i login

error:   adalAuth.logoutUser is not a function
info:    Error information has been recorded to /Users/mohit/.azure/azure.err
error:   login command failed

I removed cli and re-installed it

$ azure -v
0.9.9

**Comment by yugangw-msft** [comment]
@mohitarora, could you please delete the streamline precompile folder and see whether this works? For windows, the folder is under your user profile folder such as C:\Users\<name>\.streamline. Let me know if you are at osx, i can tell you the folder path as well.

**Comment by mohitarora** [comment]
I was indeed on OSX and deleted the folder and that helped. Thanks a lot.

**Comment by yugangw-msft** [comment]
@mohitarora glad it helped. Did you use "npm install" to get the cli or use mac installer?
We are considering to clean up the streamline folder as part of installation so to eliminate this error.

**Comment by mohitarora** [comment]
I used npm install to get the cli

**Comment by yugangw-msft** [comment]
@mohitarora, thank you, this confirms.

**Comment by jinmiaoms** [comment]
I did do that, but it simply resubmitted my form.  I thought I even tried both Edge and IE. 
Unfortunately I lost the environment that had this issue.  The new VMs I created didn't prompt me for 2FA at all, which is also strange ...

Date: Mon, 28 Sep 2015 14:11:22 -0700

**Comment by yugangw-msft** [comment]
@jinmiaoms, AAD has deployed a bug fix which should get the spinning issue go away. JFYI, with the release-0.9.10, we have put in more improvement on the device flow and will support login using non org-id

**Comment by squillace** [comment]
@yugangw-msft: Can you elaborate more on the end result here? Using azure login -u with EITHER an org id OR a msa will take us to the 2fa page, which will then auth us without requiring the SECOND button to be pushed? or, what?

**Comment by yugangw-msft** [comment]
@Squillace, the spec is here:

•	`azure login` full interactive through device code (interactive, device code, default case)
•	`azure login -u {username}` interactive w/ prompt for password (semi-interactive, non device code)
•	`azure login -u {username} -p {password}` login w/ username and password. Upon failure directed w/ “Failed to authenticate because interactive login is required. ” (non-interactive, non device code)
•	`azure login —service-principal …` to use a service principal. (non-interactive, non device code)

**Comment by squillace** [comment]
and this works with either MSA or aad id? (not discussing the SP case here)

**Comment by yugangw-msft** [comment]
@squillace, the sceanario #1 of using "azure login" w/o any parameters works for both MSA and aad id.
I have edited my previous comment

**Comment by squillace** [comment]
@yugangw-msft double confirmation: NONE of the other methods (lines 2-4, above) can be used with MSA ids? ONLY the full interactive version?

**Comment by yugangw-msft** [comment]
@squillace that is correct, because we don't invoke the new device flow which is the only way to get MSA login

**Comment by squillace** [comment]
Great. can you point me to the doc code in the project for that? I want to make sure the --help switch explains this precisely.

**Comment by squillace** [comment]
and, THANKS!!! :100:

**Comment by yugangw-msft** [comment]
@squillace, it is here (https://github.com/Azure/azure-xplat-cli/blob/release-0.9.10/lib/commands/login._js#L39). Please propose the text and i will incorporate. I know we will mention the instruction in the release note, but it is not maintainable.

**Comment by squillace** [comment]
Help should be completely in `--help`. It's the "man page" way. Can you explain this line?

`.option('-e --environment [environment]', $('Environment to authenticate against, must support active directory'))`

What is meant here by "environment"? This, unfortunately, doesn't mean anything at all without context...

**Comment by yugangw-msft** [comment]
Check out the code here(https://github.com/Azure/azure-xplat-cli/blob/dev/lib/util/profile/environment.js#L148d).
In short, this is the parameter people can use to specify other cloud env, such as ChinaCloud, and in near future, we will add US Government Cloud. You can also run "azure account env add" to add other env such as an internal test env.

**Comment by squillace** [comment]
ah! that kind of environment; perfect, thanks. I'll submit a PR, and you all can review.

**Comment by yugangw-msft** [comment]
@Squillace, thank you! Please submit the PR to dev branch, I will integrate it to the release branch if we get a chance to refresh the installer.

**Comment by squillace** [comment]
Submitted https://github.com/Azure/azure-xplat-cli/pull/2189.
